src/support/small_set.h

```diff
@@ -0,0 +1,302 @@
+    // movement in switching between fixed and flexible mode.
+    return flexible.empty();
+  }
+
+public:
+  using value_type = T;
+  using key_type = T;
+  using reference = T&;
+  using const_reference = const T&;
+  using set_type = FlexibleSet;
+  using size_type = size_t;
+
+  SmallSetBase() {}
+  SmallSetBase(std::initializer_list<T> init) {
+    for (T item : init) {
+      insert(item);
+    }
+  }
+
+  void insert(const T& x) {
+    if (usingFixed()) {
+      if (count(x)) {
+        // The item already exists, so there is nothing to do.
+        return;
+      }
+      // We must add a new item.
+      if (usedFixed < N) {
+        // Room remains in the fixed storage.
+        fixed[usedFixed++] = x;
+      } else {
+        // No fixed storage remains. Switch to flexible.
+        assert(usedFixed == N);
+        assert(flexible.empty());
+        flexible.insert(fixed.begin(), fixed.begin() + usedFixed);
+        flexible.insert(x);
+        assert(!usingFixed());
+        usedFixed = 0;
+      }
+    } else {
+      flexible.insert(x);
+    }
+  }
+
+  void erase(const T& x) {
+    if (usingFixed()) {
+      for (size_t i = 0; i < usedFixed; i++) {
+        if (fixed[i] == x) {
+          // We found the item; erase it by moving the final item to replace it
+          // and truncating the size.
+          usedFixed--;
+          fixed[i] = fixed[usedFixed];
+          return;
+        }
+      }
+    } else {
+      flexible.erase(x);
+    }
+  }
+
+  size_t count(const T& x) {
+    if (usingFixed()) {
+      // Do a linear search.
+      for (size_t i = 0; i < usedFixed; i++) {
+        if (fixed[i] == x) {
+          return 1;
+        }
+      }
+      return 0;
+    } else {
+      return flexible.count(x);
+    }
+  }
+
+  size_t size() const {
+    if (usingFixed()) {
+      return usedFixed;
+    } else {
+      return flexible.size();
+    }
+  }
+
+  bool empty() const { return size() == 0; }
+
+  void clear() {
+    usedFixed = 0;
+    flexible.clear();
+  }
+
+  bool operator==(const SmallSetBase<T, N, FlexibleSet>& other) const {
+    if (usingFixed()) {
+      if (usedFixed != other.usedFixed) {
+        return false;
+      }
+      for (size_t i = 0; i < usedFixed; i++) {
+        bool found = false;
+        for (size_t j = 0; j < usedFixed; j++) {
+          if (fixed[i] == other.fixed[j]) {
+            found = true;
+            break;
+          }
+        }
+        if (!found) {
+          return false;
+        }
+      }
+      return true;
+    }
+    return flexible == other.flexible;
+  }
+
+  bool operator!=(const SmallSetBase<T, N, FlexibleSet>& other) const {
+    return !(*this == other);
+  }
+
+  // iteration
+
+  template<typename Parent, typename Iterator, typename FlexibleIterator>
+  struct IteratorBase {
+    using iterator_category = std::forward_iterator_tag;
+    using value_type = T;
+    using difference_type = long;
+    using pointer = T*;
+    using reference = T&;
```

Comment:
I think these two should be `const T*` and `const T&` for `const_iterator`

Reply:
Thanks, done.